File: .claude/hooks/pre_compact.py
#!/usr/bin/env -S uv run --script
# /// script
# requires-python = ">=3.11"
# dependencies = [
#     "python-dotenv",
# ]
# ///

"""
Tusk Pre-Compact Hook

Automatically saves a checkpoint before compaction starts to preserve context.
This prevents loss of work progress when Claude Code compacts the conversation.

Key Features:
- Quick checkpoint save before compaction
- Non-blocking execution (returns immediately)
- Works for both auto and manual compaction
- Preserves recent work state
"""

import argparse
import json
import os
import sys
import subprocess
from pathlib import Path
from datetime import datetime

try:
    from dotenv import load_dotenv
    load_dotenv()
except ImportError:
    pass  # dotenv is optional


def quick_checkpoint(trigger_type: str) -> bool:
    """Create a quick checkpoint using Tusk server.
    
    Args:
        trigger_type: "auto" or "manual" compaction trigger
        
    Returns:
        bool: True if checkpoint was saved successfully
    """
    try:
        # Import here to avoid startup delays
        import asyncio
        sys.path.insert(0, str(Path.cwd() / "src"))
        
        from tusk.server import TuskServer
        from tusk.config import TuskConfig
        from tusk.models.checkpoint import Checkpoint
        
        async def save_checkpoint():
            # Quick config and server setup
            config = TuskConfig.from_env()
            server = TuskServer(config)
            
            # Create pre-compaction checkpoint
            description = f"Pre-compaction checkpoint ({trigger_type})"
            
            checkpoint = Checkpoint(
                workspace_id=config.current_workspace,
                description=description,
                work_context="Automatically saved before compaction to preserve context",
                session_id=f"pre_compact_{datetime.now().strftime('%Y%m%d_%H%M%S')}",
            )
            
            # Set shorter TTL for pre-compaction checkpoints (24 hours)
            checkpoint.set_ttl(24 * 60 * 60)  # 24 hours in seconds
            
            # Save checkpoint
            if server.checkpoint_storage.save(checkpoint):
                # Index for search
                server.search_engine.index_checkpoint(checkpoint)
                return True
            return False
        
        # Run async checkpoint save
        result = asyncio.run(save_checkpoint())
        return result
        
    except Exception as e:
        print(f"Error creating pre-compaction checkpoint: {e}", file=sys.stderr)
        return False


def main():
    try:
        # Parse command line arguments
        parser = argparse.ArgumentParser(description="Tusk pre-compact hook")
        parser.add_argument('--save-checkpoint', action='store_true', default=True,
                          help='Save checkpoint before compaction (default: True)')
        parser.add_argument('--verbose', action='store_true',
                          help='Print verbose output')
        args = parser.parse_args()
        
        # Read JSON input from stdin
        input_data = json.loads(sys.stdin.read())
        
        # Extract trigger type
        trigger = input_data.get('trigger', 'unknown')  # "manual" or "auto"
        session_id = input_data.get('session_id', 'unknown')
        
        if args.verbose:
            if trigger == "manual":
                print(f"📋 Manual compaction initiated (session: {session_id[:8]}...)")
            else:
                print(f"🔄 Auto-compaction triggered - context window full (session: {session_id[:8]}...)")
        
        # Save checkpoint if enabled
        if args.save_checkpoint:
            if args.verbose:
                print("📸 Saving quick checkpoint before compaction...")
            
            success = quick_checkpoint(trigger)
            
            if args.verbose:
                if success:
                    print("✅ Checkpoint saved successfully")
                else:
                    print("❌ Failed to save checkpoint")
        
        # Log the event (basic logging)
        log_dir = Path("logs")
        log_dir.mkdir(parents=True, exist_ok=True)
        
        log_entry = {
            "timestamp": datetime.now().isoformat(),
            "event": "pre_compact",
            "trigger": trigger,
            "session_id": session_id,
            "checkpoint_saved": success if args.save_checkpoint else False
        }
        
        # Append to log file
        log_file = log_dir / 'tusk_hooks.json'
        if log_file.exists():
            with open(log_file, 'r') as f:
                try:
                    logs = json.load(f)
                except json.JSONDecodeError:
                    logs = []
        else:
            logs = []
        
        logs.append(log_entry)
        
        with open(log_file, 'w') as f:
            json.dump(logs, f, indent=2)
        
        if args.verbose:
            print(f"📝 Event logged to {log_file}")
        
        # Return success (allows compaction to proceed)
        sys.exit(0)
        
    except json.JSONDecodeError:
        # Handle JSON decode errors gracefully
        print("Warning: Could not parse hook input", file=sys.stderr)
        sys.exit(0)
    except KeyboardInterrupt:
        print("Hook interrupted", file=sys.stderr)
        sys.exit(0)
    except Exception as e:
        print(f"Hook error: {e}", file=sys.stderr)
        # Don't fail compaction due to hook errors
        sys.exit(0)


if __name__ == '__main__':
    main()

File: .claude/hooks/pre_compact.ts
#!/usr/bin/env bun
/**
 * Tusk Enhanced Pre-Compact Hook
 *
 * Captures rich context before compaction including:
 * - Git status and recent changes
 * - Project context files
 * - Session metadata
 * - Transcript backup creation
 *
 * This prevents loss of work progress when Claude Code compacts the conversation.
 * Cross-platform compatible for Windows, macOS, and Linux.
 */

import { spawnSync } from "bun";
import { logSuccess, logError, findTuskCli } from "./hook-logger.ts";
import * as fs from "fs";
import * as path from "path";
import * as os from "os";

interface ClaudeCompactInput {
  session_id?: string;
  transcript_path?: string;
  trigger?: "manual" | "auto";
  custom_instructions?: string;
}

async function getGitContext(): Promise<string> {
  try {
    // Get current branch
    const branchResult = spawnSync(["git", "rev-parse", "--abbrev-ref", "HEAD"], {
      stdout: "pipe", stderr: "pipe", timeout: 5000
    });
    const branch = branchResult.success
      ? new TextDecoder().decode(branchResult.stdout).trim()
      : "unknown";

    // Get uncommitted changes
    const statusResult = spawnSync(["git", "status", "--porcelain"], {
      stdout: "pipe", stderr: "pipe", timeout: 5000
    });

    let changes = [];
    if (statusResult.success) {
      const output = new TextDecoder().decode(statusResult.stdout).trim();
      changes = output ? output.split('\n').filter(line => line.trim()) : [];
    }

    // Get recent commits (last 3)
    const logResult = spawnSync(["git", "log", "--oneline", "-3"], {
      stdout: "pipe", stderr: "pipe", timeout: 5000
    });
    const recentCommits = logResult.success
      ? new TextDecoder().decode(logResult.stdout).trim()
      : "";

    let context = `Branch: ${branch}`;
    if (changes.length > 0) {
      context += `\nUncommitted changes: ${changes.length} files`;
      context += `\n${changes.slice(0, 5).join('\n')}`;
    }
    if (recentCommits) {
      context += `\n\nRecent commits:\n${recentCommits}`;
    }

    return context;
  } catch {
    return "Git context unavailable";
  }
}

async function getProjectContext(): Promise<string> {
  const contextFiles = [
    ".claude/CONTEXT.md",
    ".claude/TODO.md",
    "TODO.md",
    "README.md",
    "CLAUDE.md"
  ];

  const contexts = [];

  for (const file of contextFiles) {
    try {
      if (fs.existsSync(file)) {
        const content = fs.readFileSync(file, 'utf-8').trim();
        if (content) {
          contexts.push(`=== ${file} ===\n${content.substring(0, 500)}${content.length > 500 ? '...' : ''}`);
        }
      }
    } catch {
      // Skip files we can't read
    }
  }

  return contexts.length > 0 ? contexts.join('\n\n') : "No project context files found";
}

async function backupTranscript(transcriptPath: string, trigger: string): Promise<string | null> {
  if (!transcriptPath || !fs.existsSync(transcriptPath)) {
    return null;
  }

  try {
    const tuskDir = path.join(os.homedir(), ".tusk");
    const backupDir = path.join(tuskDir, "transcript-backups");

    // Ensure backup directory exists
    fs.mkdirSync(backupDir, { recursive: true });

    const timestamp = new Date().toISOString().replace(/[:T]/g, '-').slice(0, 19);
    const backupPath = path.join(backupDir, `transcript-${timestamp}-${trigger}.txt`);

    // Copy transcript to backup location
    fs.copyFileSync(transcriptPath, backupPath);

    return backupPath;
  } catch {
    return null;
  }
}

async function main() {
  let claudeCodeSessionId: string | undefined;

  try {
    // Read JSON input from stdin if available
    let inputData: ClaudeCompactInput = {};
    try {
      const stdinBuffer = [];
      for await (const chunk of process.stdin) {
        stdinBuffer.push(chunk);
      }
      if (stdinBuffer.length > 0) {
        inputData = JSON.parse(Buffer.concat(stdinBuffer).toString()) || {};
        claudeCodeSessionId = (inputData as any).session_id;
      }
    } catch {
      // No input or invalid JSON - continue with empty data
    }

    const now = new Date();
    const trigger = inputData.trigger || "auto";

    // Gather rich context
    const gitContext = await getGitContext();
    const projectContext = await getProjectContext();

    // Backup transcript if available
    let transcriptBackup = null;
    if (inputData.transcript_path) {
      transcriptBackup = await backupTranscript(inputData.transcript_path, trigger);
    }

    // Create comprehensive description
    const baseDescription = `Pre-compaction context preservation (${trigger})`;
    const contextDetails = [
      "=== Session Info ===",
      `Session ID: ${inputData.session_id || 'unknown'}`,
      `Timestamp: ${now.toLocaleString()}`,
      `Trigger: ${trigger}`,
      "",
      "=== Git Context ===",
      gitContext,
      "",
      "=== Project Context ===",
      projectContext
    ];

    if (transcriptBackup) {
      contextDetails.push("", "=== Transcript Backup ===", `Saved to: ${transcriptBackup}`);
    }

    const description = `${baseDescription}\n\n${contextDetails.join('\n')}`;

    // Find tusk CLI using smart path resolution
    const cliPath = findTuskCli(import.meta.path);
    if (!cliPath) {
      logError("pre_compact", "CLI not found in any expected location");
      console.error(`⚠️ Tusk CLI not found. Set TUSK_CLI_PATH environment variable or ensure tusk is in a standard location.`);
      process.exit(0);
    }

    // Create checkpoint with rich context
    const tags = [
      "pre-compaction",
      "context-preservation",
      trigger,
      "auto-checkpoint"
    ];

    const cliArgs = [
      "bun", cliPath, "checkpoint", description,
      tags.join(","),
      "--entry-type=auto-save",
      "--confidence=0.9"
    ];

    // Add session ID if available
    if (claudeCodeSessionId) {
      cliArgs.push(`--session-id=${claudeCodeSessionId}`);
    }

    const result = spawnSync(cliArgs, {
      stdout: "pipe",
      stderr: "pipe",
    });

    if (result.success) {
      const sessionInfo = claudeCodeSessionId ? ` [${claudeCodeSessionId.slice(0, 8)}...]` : "";
      logSuccess("pre_compact", `rich context preserved (${trigger})${sessionInfo}`);
      console.error(`✅ Pre-compaction checkpoint saved`);
      console.error(`💡 Your active plan and context will survive - use recall() after restart`);
    } else {
      const errorOutput = new TextDecoder().decode(result.stderr);
      logError("pre_compact", errorOutput);
      console.error(`⚠️ Pre-compact checkpoint failed: ${errorOutput}`);
    }
  } catch (error) {
    logError("pre_compact", String(error));
    console.error(`⚠️ Pre-compact checkpoint error: ${error}`);
  }

  // Always exit successfully to not block compaction
  process.exit(0);
}

main();